rotkehlchen/tests/data_migrations/test_migration_20.py

```diff
@@ -345,3 +363,150 @@ def test_migration_20_edge_case_multiple_swaps(database: DBHandler) -> None:
             assert HistoryEventSubType.SPEND.serialize() in subtypes
             assert HistoryEventSubType.RECEIVE.serialize() in subtypes
             assert len(events) == 2
+
+
+@pytest.mark.parametrize('use_custom_database', ['botched_v47_rotkehlchen.db'])
+@pytest.mark.parametrize('perform_upgrades_at_unlock', [False])
+@pytest.mark.parametrize('data_migration_version', [19])
+def test_migration_20_recover_lost_trades(database: DBHandler) -> None:
+    """Test that migration 20 correctly recovers trades lost during v1.39.0 upgrade.
+
+    The v1.39.0 upgrade had a bug where trades with the same location and empty link field
+    would generate identical event identifiers. Since identifiers must be unique, only the
+    first trade was kept while subsequent trades were silently dropped from the database.
+
+    This test verifies that the recovery function:
+    1. Detects affected databases by looking for events with location hash identifiers
+    2. Finds and opens the v47 backup created before the problematic upgrade
+    3. Extracts trades with empty link fields from the backup
+    4. Converts them to the new SwapEvent format with proper unique event identifiers
+    5. Removes the incorrectly imported events and replaces them with recovered ones
+    """
+    rotki = MockRotkiForMigrations(database)
+    shutil.copy(
+        src=(backup_db_path := Path(__file__).resolve().parent.parent / 'data' / 'v47_rotkehlchen.db'),  # noqa: E501
+        dst=database.user_data_dir / f'{ts_now()}_rotkehlchen_db_v47.backup',
+    )
+    unlock_database(
+        db_connection=(backup_conn := DBConnection(
+            path=str(backup_db_path),
+            connection_type=DBConnectionType.USER,
+            sql_vm_instructions_cb=database.sql_vm_instructions_cb,
+        )),
+        password=database.password,
+        sqlcipher_version=database.sqlcipher_version,
+    )
+    with backup_conn.read_ctx() as cursor:
+        assert cursor.execute("SELECT timestamp, location, base_asset, quote_asset, type, amount, rate, fee, fee_currency, link, notes FROM trades WHERE link=''").fetchall() == [  # noqa: E501
+            # Two trades at the same timestamp and same location
+            (1749566127, 'A', 'ETH', 'USD', 'A', '10', '2732.36750009618', '12', 'USD', '', None),
+            (1749566127, 'A', 'ETH', 'USD', 'A', '5', '2732.36750009618', None, None, '', ''),
+            # Same location ('A') but different timestamps
+            (1749566154, 'A', 'BTC', 'USD', 'A', '1', '108659.316521703', None, None, '', ''),
+            (1749566160, 'A', 'BTC', 'USD', 'A', '2', '108659.316521703', None, None, '', ''),
```

Comment:
Also put one with link = None and see that the other part of the migration fixes it. It should right?

Reply:
> Also put one with link = None and see that the other part of the migration fixes it. It should right?

i think you're getting the migrations mixed up(or i am). the other one handles botched swap events due to `trades.unique_id` being `NULL`. so adding a trade to the test with `unique_id=None` wouldn't actually hit that migration since it's now properly handled by the db upgrade. but i went ahead and added a `None` entry anyway and it works perfectly in the upgrade from v47 to v48.